### PR TITLE
Improve error handling + consistency when no_std

### PR DESCRIPTION
**Pull request by jsdw**
I standardised no_std errors to be `Display + Debug`, which is closer to the "real" `std::error::Error` trait (and is what `serde` does too for `no_std`), exposed a couple of helpers to make constructing them from strings simpler (`scale-value` and `subxt` use these) and made use of `derive_more` since it's used in `scale-value` anyway, and is rather helpful.

Also bumped the version ready for a new release once this merges.

Part of a series:

- You are here: https://github.com/paritytech/scale-decode/pull/31
- https://github.com/paritytech/scale-encode/pull/13
- https://github.com/paritytech/scale-value/pull/40

Changes tested against Subxt; I'll update that next.

### CODE REVIEW

**Review by tadeohepperle — APPROVED** [review]
Looks good to me :)

File: scale-decode/src/error/mod.rs [hunk 1]
```diff
@@ -83,107 +104,73 @@ impl From<DecodeError> for Error {
...
-#[derive(Debug)]
+#[derive(Debug, derive_more::From, derive_more::Display)]
 pub enum ErrorKind {
     /// Something went wrong decoding the bytes based on the type
     /// and type registry provided.
+    #[from]
+    #[display(fmt = "Error decoding bytes given the type ID and registry provided: {_0}")]
     VisitorDecodeError(DecodeError),
     /// We cannot decode the number seen into the target type; it's out of range.
+    #[display(fmt = "Number {value} is out of range")]
```
Comment (gilescope) [inline scale-decode/src/error/mod.rs:115]:
```suggestion
    #[display(fmt = "Number '{value}' is out of range")]
```

File: scale-decode/src/visitor/mod.rs [hunk 2]
```diff
@@ -279,136 +278,90 @@ pub trait Visitor: Sized {
...
     /// We ran into an error trying to decode a bit sequence.
+    #[from]
+    #[display(fmt = "Cannot decode bit sequence: {_0}")]
     BitSequenceError(BitSequenceError),
     /// The type we're trying to decode is supposed to be compact encoded, but that is not possible.
+    #[display(fmt = "Could not decode compact encoded type into {_0:?}")]
     CannotDecodeCompactIntoType(scale_info::Type<PortableForm>),
     /// Failure to decode bytes into a string.
+    #[from]
+    #[display(fmt = "Could not decode string: {_0}")]
```
Comment (gilescope) [inline scale-decode/src/visitor/mod.rs:292]:
```suggestion
    #[display(fmt = "Could not decode string: '{_0}'")]
```

**Review by gilescope — APPROVED** [review]

**Review by lexnv — APPROVED** [review]

**Comment by jsdw** [comment]
I'll hold off releasing this until @tadeohepperle your fix is in, because that would be very cool to get out :)

**Pull request merged**